[WIP] Depth and IMU data processing

nadimasmar opened the pull request:
This PR restructures the sensors package and adds processing for incoming IMU and depth data.

1. Added `ImuRepublisher` and `DepthRepublisher` processors and integrated them into a single `sensor_node` executable for easier state sharing (e.g., using the latest IMU orientation when processing depth).

2. Updated CMakeLists.txt to build/install sensor_node and removed obsolete executables/sources from the package.

3. Updated launch to start the Xsens driver only when sim:=false, `sensor_node` runs regardless of sim. 

4. Updated package.xml to include Eigen dependencies and refreshed README.md with detailed IMU/depth processing documentation.

TODO:
- Add a DVL data processor. 
- decide on a sensor fusion procedure or a simple state aggregator
- Test and iterate

Stuart-Klenner committed:
Create imuprocessing

nadimasmar committed:
Write README for sensor data processing details

Expanded README to include detailed sensor data processing information, including IMU, depth sensor, and DVL processing. Added sections for usage, nodes, published and subscribed topics, installation instructions, and dependencies.

Stuart-Klenner committed:
created imuprocessing.h file

created imuprocessing.h file. Takes acceleration readings from imu sensor and derives acceleration in the pool frame of reference. Requires nadim to add helper function to create rotation matrix from quarternion input from imu.

Stuart-Klenner committed:
imu processing pipeline

nadimasmar committed:
docs: Update README with detailed IMU and depth sensor processing explanations

nadimasmar committed:
Fix formatting of mathematical equation in README

nadimasmar committed:
Imu processor, package restructuring

nadimasmar committed:
Added sensor node, and finished logic for data processors.

nadimasmar committed:
Merge remote-tracking branch 'origin/main' into dev/llmweapons

nadimasmar committed:
Enhance README with sensor_node details

Updated README.md to provide details about the sensor_node, including inputs, outputs, published and subscribed topics, dependencies, building, and running instructions.

nadimasmar committed:
fix: Update CMakeLists.txt to use Eigen3_cmake_module and make sensor_node run regardless of sim.

nadimasmar committed:
Clarify launch command for sensor data processors

Updated the README to clarify the launch command for sensors.

JPGC04 reviewed (dismissed):
Very nice progress in this pr. All this needs to be tested in real life to make sure data is actually being published in real life.

WilliamZhang2205 reviewed (commented):
Minor stuff in comments I left. Great BOOMIN work overall. I agree with what the big boss JP man say, what he said

nadimasmar committed:
Corrected topic name, small README update

nadimasmar committed:
spelling fix

nadimasmar committed:
refactor: Simplify IMU processing by using precomputed quaternion inverses

nadimasmar committed:
Updated comments for clarity

WilliamZhang2205 committed:
Corrected sensors launch file by making it more clean and organized with the new nodes. JP approved commit message uwu

nadimasmar committed:
Dev/thrust allocation (#25)

Co-authored-by: Stuart-Klenner <stuartjklenner@gmail.com>
Co-authored-by: hutchallison <aahutchings06@gmail.com>
Co-authored-by: Vincenzo Bigattini <vbigattini2004@gmail.com>
Co-authored-by: anna joy <146257353+thirteenlizards@users.noreply.github.com>
Co-authored-by: John-Paul Chouery <johnpaul.chouery@gmail.com>

nadimasmar committed:
Refactored package: Each processor is now a `Node`  + new state aggregator

nadimasmar committed:
Change parameter type

nadimasmar committed:
Merge remote-tracking branch 'origin/main' into dev/llmweapons

Stuart-Klenner committed:
Adding DVL data processing documentation

Stuart-Klenner committed:
Changed  DVL data processing documentation

Stuart-Klenner committed:
Changed  DVL data processing documentation

Stuart-Klenner committed:
Changed  DVL data processing documentation

nadimasmar committed:
Added link to external and tweaked notation  to follow Textbook

Updated the DVL processing section to clarify the transformation matrix notation and added a link to relevant documentation.

nadimasmar committed:
Corrected free acceleration calculation. Adopted +w convention in quaternions

nadimasmar committed:
Merge branch 'dev/llmweapons' of https://github.com/mcgill-robotics/AUV-2026 into dev/llmweapons

nadimasmar committed:
Corrected naming convention and IMU raw data frame of reference

nadimasmar committed:
Small fix: Class constructor

Alb33rt reviewed (dismissed):
Changes verified via real-world testing. Suggest changing the name of the quaternions because they aren't very readable for less mathematically inclined engineers.

nadimasmar committed:
refactored sensors package to follow better coding practices

nadimasmar committed:
Corrected treatment of depth as +Z position. Depth is -Z position

kaidalisohaib commented on ros2_ws/src/sensors/CMakeLists.txt:
```diff
@@ -1,43 +1,72 @@
-	include/
-	DESTINATION include/${PROJECT_NAME})
+        include/
+        DESTINATION include/${PROJECT_NAME})
```
For `launch/`, the trailing slash causes the files to be installed into `share/sensors/` instead of `share/sensors/launch/`.

For the `include` part, the headers file are being installed to `include/sensors/sensors` instead of the wanted `include/sensors`.

Something along these lines should fix those points:

```
install(DIRECTORY
        launch
        DESTINATION share/${PROJECT_NAME})

install(DIRECTORY 
        params
        DESTINATION share/${PROJECT_NAME}
)

install(DIRECTORY
        include/
        DESTINATION include)
```

nadimasmar committed:
Switched Hard-coded paths with dynamic ones.
Co-authored-by: Sohaib Kaidali  <sohaib.kaidali@mail.mcgill.ca>

nadimasmar committed:
Update README

nadimasmar committed:
Added depth calibration service.

nadimasmar committed:
Added target dependancy for state aggregator
